dexter/src/main/java/com/karumi/dexter/DexterInstance.java

```diff
@@ -72,79 +87,136 @@ void checkPermission(String permission, PermissionListener listener) {
-    switch (permissionState) {
-      case PackageManager.PERMISSION_DENIED:
-        handleDeniedPermission(permission);
-        break;
-      case PackageManager.PERMISSION_GRANTED:
-      default:
-        finishWithGrantedPermission(permission);
-        break;
+    Collection<String> deniedRequests = new LinkedList<>();
+    Collection<String> grantedRequests = new LinkedList<>();
+
+    for (String permission : pendingPermissions) {
+      int permissionState = androidPermissionService.checkSelfPermission(activity, permission);
+      switch (permissionState) {
+        case PackageManager.PERMISSION_DENIED:
+          deniedRequests.add(permission);
+          break;
+        case PackageManager.PERMISSION_GRANTED:
+        default:
+          grantedRequests.add(permission);
+          break;
+      }
     }
+
+    handleDeniedPermissions(deniedRequests);
+    updatePermissionsAsGranted(grantedRequests);
   }
 
   /**
-   * Method called whenever the permission has been granted by the user
+   * Method called whenever the permissions has been granted by the user
    */
-  void onPermissionRequestGranted() {
-    finishWithGrantedPermission(permission);
+  void onPermissionRequestGranted(Collection<String> permissions) {
+    updatePermissionsAsGranted(permissions);
   }
 
   /**
-   * Method called whenever the permission has been denied by the user
+   * Method called whenever the permissions has been denied by the user
    */
-  void onPermissionRequestDenied() {
-    finishWithDeniedPermission(permission);
+  void onPermissionRequestDenied(Collection<String> permissions) {
+    updatePermissionsAsDenied(permissions);
   }
 
   /**
-   * Method called when the user has been informed with the rationale and agrees to continue
+   * Method called when the user has been informed with a rationale and agrees to continue
    * with the permission request process
    */
-  void onContinuePermissionRequest(String permission) {
-    requestPermission(permission);
+  void onContinuePermissionRequest() {
+    requestPermissionsToSystem(pendingPermissions);
   }
 
   /**
-   * Method called when the user has been informed with the rationale and decides to cancel
+   * Method called when the user has been informed with a rationale and decides to cancel
    * the permission request process
    */
-  void onCancelPermissionRequest(String permission) {
-    finishWithDeniedPermission(permission);
+  void onCancelPermissionRequest() {
+    updatePermissionsAsDenied(pendingPermissions);
   }
 
   /**
-   * Starts the native request permission process
+   * Starts the native request permissions process
    */
-  void requestPermission(String permission) {
-    int permissionCode = getPermissionCodeForPermission(permission);
-    ActivityCompat.requestPermissions(activity, new String[]{permission}, permissionCode);
+  void requestPermissionsToSystem(Collection<String> permissions) {
+    int requestCode = getRequestCodeForPermissions(permissions);
```

Comment:
This could be a constant ;) The ``requestCode`` used is not going to affect the execution result.